{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "302f61a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Remove baseline from amplification curves (BETA)\n",
    "#'\n",
    "#' Remove baseline from qPCR amplification curves\n",
    "#' by subtracting median of initial cycles.\n",
    "#'\n",
    "#' BETA function version because:\n",
    "#'\n",
    "#' - assumes Roche Lightcycler format,\n",
    "#' we should ideally replace \"program_no == 2\" by something more generic?\n",
    "#'\n",
    "#' - the rule-of thumb \"baseline is median of initial 10 cycles\"\n",
    "#' has not been tested robustly\n",
    "#'\n",
    "#' @param plateamp data frame with plate amplification data, including variables\n",
    "#'   well, cycle, fluor_raw (raw fluorescence value), and program_no. Assume\n",
    "#'   program 2 for amplification curves from Roche Lightcycler format data.\n",
    "#' @param maxcycle maximum cycle value to use for baseline, before\n",
    "#'   amplification.\n",
    "#'\n",
    "#' @return platemap with additional columns per well:\n",
    "#' \\tabular{ll}{\n",
    "#'   fluor_base      \\tab baseline /background value  \\cr\n",
    "#'   fluor_signal    \\tab normalized fluorescence signal, \n",
    "#'   i.e. fluor_raw - fluor_base\n",
    "#'   }\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tidyr %>%\n",
    "#' @importFrom rlang .data\n",
    "#'\n",
    "#' @examples\n",
    "#' # create simple dataset of raw fluorescence\n",
    "#' # with two samples over 15 cycles\n",
    "#' raw_fluor_tibble <- tibble(sample_id = rep(c(\"S1\", \"S2\"), each = 15),\n",
    "#'                           target_id = \"T1\",\n",
    "#'                           well_row = \"A\",\n",
    "#'                           well_col = rep(c(1, 2), each = 15),\n",
    "#'                           well = rep(c(\"A1\", \"A2\"), each = 15),\n",
    "#'                           cycle = rep(1:15,2),\n",
    "#'                           fluor_raw = c(1:15, 6:20),\n",
    "#'                           program_no = 2)\n",
    "#'\n",
    "#' # remove base fluorescence from dataset\n",
    "#' raw_fluor_tibble %>%\n",
    "#'     debaseline()\n",
    "#'     \n",
    "debaseline <- function(plateamp, maxcycle = 10) {\n",
    "    assertthat::assert_that(\n",
    "        assertthat::has_name(\n",
    "            plateamp,\n",
    "            c(\"well\", \"program_no\", \"cycle\", \"fluor_raw\"))\n",
    "    )\n",
    "    assertthat::assert_that(\n",
    "        !assertthat::has_name(plateamp,\"fluor_base\"),\n",
    "        msg = \"plateamp contains a variable called fluor_base, which breaks debaseline\"\n",
    "    )\n",
    "    baseline <-\n",
    "        plateamp %>%\n",
    "        dplyr::group_by(.data$well) %>%\n",
    "        dplyr::filter(.data$program_no == 2,\n",
    "                      .data$cycle <= maxcycle) %>%\n",
    "        dplyr::summarize(fluor_base = stats::median(.data$fluor_raw))\n",
    "    plateamp %>%\n",
    "        dplyr::left_join(baseline, by = \"well\") %>%\n",
    "        dplyr::mutate(fluor_signal = .data$fluor_raw - .data$fluor_base)\n",
    "}\n",
    "\n",
    "\n",
    "#' Calculate dy/dx vector from vectors y and x\n",
    "#'\n",
    "#' Used in tidyqpcr to calculate dR/dT for a melt curve of fluorescence signal R\n",
    "#' vs temperature T.\n",
    "#'\n",
    "#' @param x input variable, numeric vector, assumed to be temperature\n",
    "#' @param y output variable, numeric vector of same length as x, assumed\n",
    "#'   to be fluorescence signal.\n",
    "#' @param method to use for smoothing:\n",
    "#'\n",
    "#'   \"spline\" default, uses smoothing spline stats::smooth.spline.\n",
    "#'\n",
    "#'   \"diff\" base::diff for lagged difference\n",
    "#'\n",
    "#' @param ... other arguments to pass to smoothing method.\n",
    "#'\n",
    "#' @return estimated first derivative of y with respect to x, numeric vector of\n",
    "#'   same length as y.\n",
    "#'\n",
    "#' @family melt_curve_functions\n",
    "#'\n",
    "#' @export\n",
    "#' \n",
    "#' @examples\n",
    "#' # create simple curve\n",
    "#' x = 1:5\n",
    "#' y = x^2\n",
    "#'\n",
    "#' # calculate gradient of curve\n",
    "#' #----- use case 1 : using splines\n",
    "#' calculate_dydx(x, y)\n",
    "#' \n",
    "#' # optional arguments are passed to smooth.splines function\n",
    "#' calculate_dydx(x, y, spar = 0.5)\n",
    "#' \n",
    "#' #----- use case 2 : using difference between adjacent points\n",
    "#' calculate_dydx(x, y, method = \"diff\")\n",
    "#'\n",
    "calculate_dydx <- function(x, y, method = \"spline\", ...) {\n",
    "    assertthat::assert_that(is.numeric(x))\n",
    "    assertthat::assert_that(is.numeric(y))\n",
    "    assertthat::assert_that(length(x) == length(y))\n",
    "    if (method == \"diff\") {\n",
    "       return(-c(diff(y) / diff(x), NA))\n",
    "    } else if (method == \"spline\") {\n",
    "        fit <- stats::smooth.spline(x = x, y = y, ...)\n",
    "        return(-1 * stats::predict(object = fit, x = x, deriv = 1)$y)\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "#' Calculate dR/dT of melt curves for of every well in a plate.\n",
    "#'\n",
    "#' dR/dT, the derivative of the melt curve (of fluorescence signal R vs\n",
    "#' temperature T), has a maximum at the melting temperature Tm. A single peak in\n",
    "#' this suggests a single-length PCR product is present in the well.\n",
    "#' \n",
    "#' Note that this function does not group by plate, only by well.\n",
    "#' The function will give strange results if you pass it data from \n",
    "#' more than one plate. Avoid this by analysing one plate at a time.\n",
    "#'\n",
    "#' @param platemelt data frame describing melt curves, including variables\n",
    "#'   well, temperature, fluor_raw (raw fluorescence value).\n",
    "#' @param method to use for smoothing:\n",
    "#'\n",
    "#'   \"spline\" default, uses smoothing spline stats::smooth.spline.\n",
    "#'\n",
    "#'   \"diff\" base::diff for lagged difference\n",
    "#'\n",
    "#' @param ... other arguments to pass to smoothing method.\n",
    "#'\n",
    "#' @return platemelt with additional column dRdT.\n",
    "#'\n",
    "#' @family melt_curve_functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tidyr %>%\n",
    "#'\n",
    "#' @examples\n",
    "#' # create simple curve\n",
    "#' # create simple dataset of raw fluorescence with two samples\n",
    "#' temp_tibble <- tibble(sample_id = rep(c(\"S1\", \"S2\"), each = 10),\n",
    "#'                           target_id = \"T1\",\n",
    "#'                           well_row = \"A\",\n",
    "#'                           well_col = rep(c(1, 2), each = 10),\n",
    "#'                           well = rep(c(\"A1\", \"A2\"), each = 10),\n",
    "#'                           temperature = rep(56:65,2),\n",
    "#'                           fluor_raw = c(1:10, 6:15))\n",
    "#'\n",
    "#' # calculate drdt of all melt curves\n",
    "#' #----- use case 1 : using splines\n",
    "#' temp_tibble %>%\n",
    "#'     calculate_drdt_plate()\n",
    "#' \n",
    "#' # optional arguments are passed to smooth.splines function\n",
    "#' temp_tibble %>%\n",
    "#'     calculate_drdt_plate(spar = 0.5)\n",
    "#' \n",
    "#' #----- use case 2 : using difference between adjacent points\n",
    "#' temp_tibble %>%\n",
    "#'     calculate_drdt_plate(method = \"diff\")\n",
    "#'\n",
    "calculate_drdt_plate <- function(platemelt, method = \"spline\", ...) {\n",
    "    assertthat::assert_that(\n",
    "        assertthat::has_name(\n",
    "            platemelt,\n",
    "            c(\"well\", \"temperature\", \"fluor_raw\"))\n",
    "    )\n",
    "    if (assertthat::has_name(platemelt,\"plate\") ) {\n",
    "        warning(\"platemelt has a plate column, but calculate_drdt_plate works only for single plates.\")\n",
    "    }\n",
    "    platemelt %>%\n",
    "        dplyr::arrange(.data$well, .data$temperature) %>%\n",
    "        dplyr::group_by(.data$well) %>%\n",
    "        dplyr::mutate(dRdT =\n",
    "                          calculate_dydx(x = .data$temperature,\n",
    "                                         y = .data$fluor_raw,\n",
    "                                         method = method,\n",
    "                                         ...)\n",
    "        ) %>%\n",
    "        dplyr::ungroup()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afb39b86",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Calculate a normalized value for a subset of reference ids\n",
    "#'\n",
    "#' This is used to calculate the normalized `cq` values for reference\n",
    "#' `target_ids` (e.g. genes), to use in `delta_cq` calculation for each\n",
    "#' `sample_id`.\n",
    "#'\n",
    "#' Also used to calculate the normalized `delta_cq` values for reference\n",
    "#' `sample_ids`, to use in `deltadelta_cq` calculation for each `target_id`.\n",
    "#'\n",
    "#' @param value_df data frame containing relevant columns, those named in\n",
    "#'   `value_name` and `id_name` parameters.\n",
    "#' @param ref_ids values of reference ids, that are used to calculate\n",
    "#'   normalized reference value.\n",
    "#' @param value_name name of column containing values. This column should be\n",
    "#'   numeric.\n",
    "#' @param id_name name of column containing ids.\n",
    "#' @param norm_function Function to use to calculate the value to normalize by.\n",
    "#'   Default function is median, alternatively could use mean, geometric mean,\n",
    "#'   etc.\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tidyr %>%\n",
    "#' @importFrom stats median\n",
    "#'\n",
    "#' @examples\n",
    "#' # create simple cq dataset with one sample, two targets  and 3 reps\n",
    "#' cq_tibble <- tibble(sample_id = \"S_1\",\n",
    "#'                      target_id = rep(c(\"T_1\",\n",
    "#'                                        \"T_norm\"), each = 3),\n",
    "#'                      tech_rep = rep(1:3, 2),\n",
    "#'                      well_row = rep(c(\"A\",\n",
    "#'                                       \"B\"), each = 3),\n",
    "#'                      well_col = 1,\n",
    "#'                      well = paste0(well_row, well_col),\n",
    "#'                      cq = c(10, 10, 10,\n",
    "#'                             12, 12, 11))\n",
    "#'                      \n",
    "#' # normalise cq to reference target_id called 'T_norm'\n",
    "#' \n",
    "#' #----- use case 1: median reference target_id value\n",
    "#' cq_tibble %>%\n",
    "#'     calculate_normvalue(ref_ids = \"T_norm\",\n",
    "#'                         value_name = \"cq\",\n",
    "#'                         id_name = \"target_id\")\n",
    "#' \n",
    "#' #----- use case 2: mean reference target_id value \n",
    "#' cq_tibble %>%\n",
    "#'     calculate_normvalue(ref_ids = \"T_norm\",\n",
    "#'                         value_name = \"cq\",\n",
    "#'                         id_name = \"target_id\",\n",
    "#'                         norm_function = mean)\n",
    "#'\n",
    "calculate_normvalue <- function(value_df,\n",
    "                      ref_ids,\n",
    "                      value_name = \"value\",\n",
    "                      id_name = \"id\",\n",
    "                      norm_function = median) {\n",
    "    \n",
    "    assertthat::assert_that(\n",
    "        assertthat::has_name(value_df, \n",
    "                             c(value_name,id_name)))\n",
    "    \n",
    "    # make subset of value_df where gene is one or more ref_ids\n",
    "    value_to_norm_by <- dplyr::filter(value_df,\n",
    "                             .data[[id_name]] %in% ref_ids) %>%\n",
    "        dplyr::pull({{value_name}}) %>%\n",
    "        norm_function(na.rm = TRUE)\n",
    "    \n",
    "    # assign summary (median) value to value_df$value_to_norm_by\n",
    "    dplyr::mutate(value_df, value_to_norm_by = value_to_norm_by)\n",
    "}\n",
    "\n",
    "#' Calculate delta cq (\\eqn{\\Delta Cq}) to normalize quantification cycle\n",
    "#' (log2-fold) data within sample_id.\n",
    "#'\n",
    "#' This function implements relative quantification by the delta Cq method. For\n",
    "#' each sample, the Cq values of all targets (e.g. genes, probes, primer sets)\n",
    "#' are compared to one or more reference target ids specified in\n",
    "#' `ref_target_ids`.\n",
    "#'\n",
    "#' @param cq_df a data frame containing columns `sample_id`, value_name (default\n",
    "#'   `cq`) and tid_name (default `target_id`). Crucially, sample_id should be\n",
    "#'   the same for different technical replicates measuring identical reactions\n",
    "#'   in different wells of the plate, but differ for different biological and\n",
    "#'   experimental replicates. See tidyqpcr vignettes for examples.\n",
    "#' @param ref_target_ids names of targets to normalize by, i.e. reference\n",
    "#'   genes, hydrolysis probes, or primer sets. This can be one reference target\n",
    "#'   id, a selection of multiple target ids, or even all measured target ids. In\n",
    "#'   the case of all of them, the delta Cq value would be calculated relative to\n",
    "#'   the median (or other `norm_function`) of all measured targets.\n",
    "#' @param norm_function Function to use to calculate the value to normalize by\n",
    "#'   on given scale. Default is median, alternatively could use mean.\n",
    "#'\n",
    "#' @return data frame like cq_df with three additional columns:\n",
    "#'\n",
    "#'   \\tabular{ll}{ ref_cq    \\tab summary (median/mean) cq value for reference\n",
    "#'   target ids \\cr delta_cq  \\tab normalized value, \\eqn{\\Delta Cq} \\cr\n",
    "#'   rel_abund \\tab normalized ratio, \\eqn{2^(-\\Delta Cq)} }\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tidyr %>%\n",
    "#' @importFrom stats median\n",
    "#' @importFrom rlang .data\n",
    "#' \n",
    "#' @examples\n",
    "#' # create simple cq dataset with two samples, two targets  and 3 reps\n",
    "#' cq_tibble <- tibble(sample_id = rep(c(\"S_1\",\"S_1\",\"S_1\", \"S_2\", \"S_2\", \"S_2\"), 2),\n",
    "#'                      target_id = rep(c(\"T_1\",\n",
    "#'                                        \"T_norm\"), each = 6),\n",
    "#'                      tech_rep = rep(1:3, 4),\n",
    "#'                      well_row = rep(c(\"A\",\n",
    "#'                                       \"B\"), each = 6),\n",
    "#'                      well_col = rep(1:6, 2),\n",
    "#'                      well = paste0(well_row, well_col),\n",
    "#'                      cq = c(10, 10, 10, 12, 12, 11,\n",
    "#'                              9,  9,  9,  9,  9,  9))\n",
    "#'                      \n",
    "#' # calculate deltacq using reference target_id called 'T_norm'\n",
    "#' \n",
    "#' #----- use case 1: median reference target_id value\n",
    "#' cq_tibble %>%\n",
    "#'     calculate_deltacq_bysampleid(ref_target_ids = \"T_norm\")\n",
    "#' \n",
    "#' #----- use case 2: mean reference target_id value \n",
    "#' cq_tibble %>%\n",
    "#'     calculate_deltacq_bysampleid(ref_target_ids = \"T_norm\",\n",
    "#'                                  norm_function = mean)\n",
    "#'\n",
    "calculate_deltacq_bysampleid <- function(cq_df,\n",
    "                                         ref_target_ids,\n",
    "                                         norm_function = median) {\n",
    "    \n",
    "    assertthat::assert_that(\n",
    "        assertthat::has_name(cq_df, \n",
    "                             c(\"target_id\", \"sample_id\",\"cq\")))\n",
    "    \n",
    "    cq_df %>%\n",
    "        dplyr::group_by(.data$sample_id) %>%\n",
    "        dplyr::do(calculate_normvalue(.data,\n",
    "                                   ref_ids = ref_target_ids,\n",
    "                                   value_name = \"cq\",\n",
    "                                   id_name = \"target_id\",\n",
    "                                   norm_function = norm_function)) %>%\n",
    "        dplyr::rename(ref_cq = .data$value_to_norm_by) %>%\n",
    "        dplyr::ungroup() %>%\n",
    "        dplyr::mutate(\n",
    "               delta_cq    = .data$cq - .data$ref_cq,\n",
    "               rel_abund   = 2^-.data$delta_cq)\n",
    "}\n",
    "\n",
    "\n",
    "#' Calculate delta delta cq (\\eqn{\\Delta \\Delta Cq}) to globally normalize\n",
    "#' quantification cycle (log2-fold) data across sample_id.\n",
    "#'\n",
    "#' By default, \\eqn{\\Delta \\Delta Cq} is positive if a target is more highly\n",
    "#' detected in the relevant sample, compared to reference samples. This can be\n",
    "#' flipped by setting the parameter `ddcq_positive` to `FALSE`. In either case,\n",
    "#' The fold change, \\eqn{2^{\\Delta \\Delta Cq}}, is also reported.\n",
    "#'\n",
    "#' This function does a global normalization, where all samples are compared to\n",
    "#' one or more reference samples specified in `ref_sample_ids`. There are other\n",
    "#' experimental designs that require comparing samples in pairs or small groups,\n",
    "#' e.g. a time course comparing `delta_cq` values against a reference strain at\n",
    "#' each time point. For those situations, instead we recommend adapting code\n",
    "#' from this function, changing the grouping variables used in to\n",
    "#' `dplyr::group_by` to draw the contrasts appropriate for the experiment.\n",
    "#'\n",
    "#' @param deltacq_df a data frame containing columns `sample_id`, value_name\n",
    "#'   (default `delta_cq`) and tid_name (default `target_id`). Crucially,\n",
    "#'   sample_id should be the same for different technical replicates measuring\n",
    "#'   identical reactions in different wells of the plate, but differ for\n",
    "#'   different biological and experimental replicates.\n",
    "#'\n",
    "#'   Usually this will be a data frame that was output by\n",
    "#'   `calculate_deltacq_bysampleid`.\n",
    "#'\n",
    "#' @param ref_sample_ids reference sample_ids to normalize by\n",
    "#' @param norm_function Function to use to calculate the value to normalize by\n",
    "#'   on given scale. Default is median, alternatively could use mean.\n",
    "#' @param ddcq_positive (default TRUE) output \\eqn{\\Delta \\Delta Cq} as positive\n",
    "#'   if a target is more highly detected in the relevant sample, compared to\n",
    "#'   reference samples.\n",
    "#'\n",
    "#' @return data frame like cq_df with three additional columns:\n",
    "#'\n",
    "#'   \\tabular{ll}{ ref_delta_cq  \\tab summary (median/mean) \\eqn{\\Delta Cq}\n",
    "#'   value for target_id in reference sample ids \\cr deltadelta_cq \\tab the\n",
    "#'   normalized value, \\eqn{\\Delta \\Delta Cq} \\cr fold_change   \\tab the\n",
    "#'   normalized fold-change ratio, \\eqn{2^(-\\Delta \\Delta Cq)} }\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tidyr %>%\n",
    "#' @importFrom stats median\n",
    "#'\n",
    "#' @examples\n",
    "#' # create simple deltacq dataset with two samples, two targets and 3 reps\n",
    "#' deltacq_tibble <- tibble(sample_id = rep(c(\"S_1\",\"S_1\",\"S_1\", \"S_norm\", \"S_norm\", \"S_norm\"), 2),\n",
    "#'                      target_id = rep(c(\"T_1\",\n",
    "#'                                        \"T_2\"), each = 6),\n",
    "#'                      tech_rep = rep(1:3, 4),\n",
    "#'                      well_row = rep(c(\"A\",\n",
    "#'                                       \"B\"), each = 6),\n",
    "#'                      well_col = rep(1:6, 2),\n",
    "#'                      well = paste0(well_row,well_col),\n",
    "#'                      delta_cq = c(1, 1, 1, 3, 3, 2,\n",
    "#'                                   4, 5, 4, 5, 5, 5))\n",
    "#'                      \n",
    "#' # calculate deltadeltacq using reference target_id called 'S_norm'\n",
    "#' \n",
    "#' #----- use case 1: median reference sample_id value\n",
    "#' deltacq_tibble %>%\n",
    "#'     calculate_deltadeltacq_bytargetid(ref_sample_ids = \"S_norm\")\n",
    "#' \n",
    "#' #----- use case 2: mean reference sample_id value \n",
    "#' deltacq_tibble %>%\n",
    "#'     calculate_deltadeltacq_bytargetid(ref_sample_ids = \"S_norm\",\n",
    "#'                                  norm_function = mean)\n",
    "#'\n",
    "calculate_deltadeltacq_bytargetid <- function(deltacq_df,\n",
    "                                         ref_sample_ids,\n",
    "                                         norm_function = median,\n",
    "                                         ddcq_positive = TRUE) {\n",
    "    \n",
    "    assertthat::assert_that(\n",
    "        assertthat::has_name(deltacq_df, \n",
    "                             c(\"target_id\", \"sample_id\",\"delta_cq\")))\n",
    "    \n",
    "    ddcq_factor <- (-1) ^ ddcq_positive\n",
    "    \n",
    "    deltacq_df %>%\n",
    "        dplyr::group_by(.data$target_id) %>%\n",
    "        dplyr::do(calculate_normvalue(.data,\n",
    "                                   ref_ids = ref_sample_ids,\n",
    "                                   value_name = \"delta_cq\",\n",
    "                                   id_name = \"sample_id\",\n",
    "                                   norm_function = norm_function)) %>%\n",
    "        dplyr::rename(ref_delta_cq = .data$value_to_norm_by) %>%\n",
    "        dplyr::ungroup() %>%\n",
    "        dplyr::mutate(\n",
    "               deltadelta_cq = ddcq_factor *\n",
    "                   (.data$delta_cq - .data$ref_delta_cq),\n",
    "               fold_change   = 2 ^ .data$deltadelta_cq)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab5a28f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Calibrate primer sets / probes by calculating detection efficiency and\n",
    "#' R squared\n",
    "#'\n",
    "#' Note efficiency is given in ratio, not per cent; multiply by 100 for that.\n",
    "#'\n",
    "#' @param cq_df_1 data frame with cq (quantification cycle) data,\n",
    "#' 1 row per well.\n",
    "#'\n",
    "#' Must have columns cq, dilution.\n",
    "#'\n",
    "#' Assumes data are only for 1 probe/primer set/target_id, i.e. all values in\n",
    "#' cq_df_1 are fit with the same slope.\n",
    "#'\n",
    "#' @param formula formula to use for log-log regression fit.\n",
    "#'\n",
    "#' Default value assumes multiple biological replicates,\n",
    "#' cq ~ log2(dilution) + biol_rep.\n",
    "#'\n",
    "#' If only a single Biological Replicate, change to cq ~ log2(dilution).\n",
    "#'\n",
    "#' @return data frame with 1 single row, and columns:\n",
    "#' efficiency, efficiency.sd, r.squared.\n",
    "#'\n",
    "#' @seealso calculate_efficiency_bytargetid\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tibble tibble\n",
    "#' \n",
    "#' @examples\n",
    "#' # create simple dilution dataset\n",
    "#' dilution_tibble <- tibble(dilution = rep(c(1, 0.1, 0.001, 0.0001), 2),\n",
    "#'                      cq = c(1, 3, 4, 6,\n",
    "#'                             4, 5, 6, 7),\n",
    "#'                      biol_rep = rep(c(1,2), each = 4),\n",
    "#'                      target_id = \"T1\")\n",
    "#'                      \n",
    "#' # calculate primer efficiency\n",
    "#' \n",
    "#' #----- use case 1: include difference across replicates in model\n",
    "#' dilution_tibble %>%\n",
    "#'     calculate_efficiency()\n",
    "#' \n",
    "#' #----- use case 2: ignore difference across replicates\n",
    "#' dilution_tibble %>%\n",
    "#'     calculate_efficiency(formula = cq ~ log2(dilution))\n",
    "#'\n",
    "calculate_efficiency <- function(cq_df_1, formula = cq ~ log2(dilution) + biol_rep) {\n",
    "    if (length(unique(cq_df_1$target_id)) > 1) {\n",
    "            warning(\"multiple target_ids, did you mean calculate_efficiency_bytargetid?\")\n",
    "    }\n",
    "    slopefit <- stats::lm(formula = formula, data = cq_df_1)\n",
    "    slopefitsummary <- summary(slopefit)\n",
    "    tibble(efficiency = -slopefit$coefficients[2],\n",
    "           efficiency.sd = slopefitsummary$coefficients[2, 2],\n",
    "           r.squared = slopefitsummary$r.squared)\n",
    "}\n",
    "\n",
    "#' Calibrate multiple probes by calculating detection efficiency and R squared\n",
    "#'\n",
    "#' See calibration vignette for example of usage.\n",
    "#'\n",
    "#' Note efficiency is given in ratio, not per cent; multiply by 100 for that.\n",
    "#'\n",
    "#' @param cq_df a data frame with cq (quantification cycle) data, 1 row per well\n",
    "#'\n",
    "#' Must have columns prep_type, target_id, cq, dilution.\n",
    "#' Only prep_type==\"+RT\" columns are used.\n",
    "#'\n",
    "#' @param formula formula to use for log-log regression fit.\n",
    "#'\n",
    "#' Default value assumes multiple biological replicates,\n",
    "#' cq ~ log2(dilution) + biol_rep.\n",
    "#'\n",
    "#' If only a single Biological Replicate, change to cq ~ log2(dilution).\n",
    "#' If multiple sample_ids, change to cq ~ log2(dilution) + sample_id.\n",
    "#'\n",
    "#' See ?formula for background and help.\n",
    "#'\n",
    "#' @param use_prep_types prep_type column values to use, default \"+RT\" for RT-qPCR.\n",
    "#'\n",
    "#' By default, this includes only reverse-transcribed values in the efficiency\n",
    "#' estimation, so excludes negative controls such as no-template and no-RT.\n",
    "#'\n",
    "#' To skip this filtering step, set use_prep_types=NA.\n",
    "#'\n",
    "#'\n",
    "#' @return data frame with columns:\n",
    "#' target_id, efficiency, efficiency.sd, r.squared.\n",
    "#'\n",
    "#' @seealso calculate_efficiency\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tidyr %>%\n",
    "#' @importFrom rlang .data\n",
    "#'\n",
    "#' @examples\n",
    "#' \n",
    "#' # create simple dilution dataset for two target_ids with two biological reps\n",
    "#' dilution_tibble <- tibble(target_id = rep(c(\"T_1\",\n",
    "#'                                             \"T_2\"), each = 8),\n",
    "#'                           well_row = rep(c(\"A\",\n",
    "#'                                            \"B\"), each = 8),\n",
    "#'                           well_col = rep(1:8, 2),\n",
    "#'                           well = paste0(well_row, well_col),\n",
    "#'                           dilution = rep(c(1, 0.1, 0.001, 0.0001), 4),\n",
    "#'                           cq = c(1, 3, 4, 6, 1, 3, 5, 7,\n",
    "#'                                  4, 5, 6, 7, 3, 7, 8, 9),\n",
    "#'                           biol_rep = rep(c(1, 1, 1, 1, 2, 2, 2, 2), 2),\n",
    "#'                           prep_type = \"+RT\")\n",
    "#'                      \n",
    "#' # calculate primer efficiency for multiple targets\n",
    "#' \n",
    "#' #----- use case 1: include difference across replicates in model\n",
    "#' dilution_tibble %>%\n",
    "#'     calculate_efficiency_bytargetid()\n",
    "#' \n",
    "#' #----- use case 2: ignore difference across replicates\n",
    "#' dilution_tibble %>%\n",
    "#'     calculate_efficiency_bytargetid(formula = cq ~ log2(dilution))\n",
    "#'\n",
    "calculate_efficiency_bytargetid <- function(cq_df,\n",
    "                           formula = cq ~ log2(dilution) + biol_rep,\n",
    "                           use_prep_types=\"+RT\") {\n",
    "    assertthat::assert_that(assertthat::has_name(cq_df, \"target_id\"))\n",
    "    \n",
    "    if (!is.na(use_prep_types)) {\n",
    "        assertthat::assert_that(assertthat::has_name(cq_df, \"prep_type\"))\n",
    "        cq_df <- dplyr::filter(cq_df, .data$prep_type %in% use_prep_types)\n",
    "    }\n",
    "    cq_df %>%\n",
    "        dplyr::group_by(.data$target_id) %>%\n",
    "        dplyr::do(calculate_efficiency(.data, formula = formula)) %>%\n",
    "        dplyr::ungroup()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ca4c540",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Create a blank plate template as a tibble (with helper functions for common plate sizes)\n",
    "#'\n",
    "#' For more help, examples and explanations, see the plate setup vignette:\n",
    "#' \\code{vignette(\"platesetup_vignette\", package = \"tidyqpcr\")}\n",
    "#'\n",
    "#' @param well_row Vector of Row labels, usually LETTERS\n",
    "#' @param well_col Vector of Column labels, usually numbers\n",
    "#' @return tibble (data frame) with columns well_row, well_col, well. This\n",
    "#'   contains all pairwise combinations of well_row and well_col, as well as\n",
    "#'   individual well names. Both well_row and well_col are coerced to factors\n",
    "#'   (even if well_col is supplied as numbers), to ensure order is consistent.\n",
    "#'\n",
    "#'   However, well is a character vector as that is the default behaviour of\n",
    "#'   \"unite\", and display order doesn't matter.\n",
    "#'\n",
    "#'   Default value describes a full 384-well plate.\n",
    "#'\n",
    "#' @examples\n",
    "#' create_blank_plate(well_row=LETTERS[1:2],well_col=1:3)\n",
    "#' \n",
    "#' create_blank_plate_96well()\n",
    "#' \n",
    "#' create_blank_plate_1536well()\n",
    "#' \n",
    "#' # create blank 96-well plate with empty edge wells\n",
    "#' \n",
    "#' create_blank_plate(well_row=LETTERS[2:7], well_col=2:11)\n",
    "#' \n",
    "#' # create blank 1536-well plate with empty edge wells\n",
    "#' \n",
    "#' full_plate_row_names <- make_row_names_lc1536()\n",
    "#' \n",
    "#' create_blank_plate(well_row=full_plate_row_names[2:31], well_col=2:47)\n",
    "#'\n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tibble tibble as_tibble\n",
    "#' @importFrom tidyr %>%\n",
    "#' @importFrom forcats as_factor\n",
    "#' @importFrom rlang .data\n",
    "#'\n",
    "create_blank_plate <- function(well_row = LETTERS[1:16], well_col = 1:24) {\n",
    "    tidyr::crossing(well_row = as_factor(well_row),\n",
    "                    well_col = as_factor(well_col)) %>%\n",
    "        as_tibble() %>%\n",
    "        tidyr::unite(\"well\", .data$well_row, .data$well_col, \n",
    "                     sep = \"\", remove = FALSE)\n",
    "}\n",
    "\n",
    "#' @describeIn create_blank_plate create blank 96-well plate\n",
    "#' @export\n",
    "#'\n",
    "create_blank_plate_96well <- function() {\n",
    "    create_blank_plate(well_row = LETTERS[1:8], well_col = 1:12)\n",
    "}\n",
    "\n",
    "#' Generates row names for the Roche Lightcycler (tm) 1536-well plates\n",
    "#' \n",
    "#' Creates a vector containing 36 row names according to the labelling system \n",
    "#' used by the Roche Lightcycler (tm)\n",
    "#' \n",
    "#' @return Vector of row names: Aa,Ab,Ac,Ad,Ba,...,Hd.\n",
    "#' \n",
    "#' @family plate creation functions\n",
    "#' \n",
    "#' @examples\n",
    "#' make_row_names_lc1536()\n",
    "#' \n",
    "#' @export\n",
    "#'\n",
    "make_row_names_lc1536 <- function() {\n",
    "    paste0(rep(LETTERS[1:8], each = 4), letters[1:4])\n",
    "}\n",
    "\n",
    "#' Generates row names for the Labcyte Echo 1536-well plates \n",
    "#' \n",
    "#' Creates a vector containing 36 row names according to the \n",
    "#' labelling system used by the Labcyte Echo\n",
    "#' \n",
    "#' @return Vector of row names: A,B,...,Z,AA,AB,...,AF.\n",
    "#' \n",
    "#' @family plate creation functions\n",
    "#' \n",
    "#' @examples\n",
    "#' make_row_names_echo1536()\n",
    "#' \n",
    "#' @export\n",
    "#'\n",
    "make_row_names_echo1536 <- function() {\n",
    "    c(LETTERS[1:26], paste0(\"A\", LETTERS[1:6]))\n",
    "}\n",
    "\n",
    "#' @describeIn create_blank_plate create blank 1536-well plate\n",
    "#' @export\n",
    "#'\n",
    "create_blank_plate_1536well <- function(\n",
    "    well_row = make_row_names_lc1536(),\n",
    "    well_col = 1:48) {\n",
    "    create_blank_plate(well_row, well_col)\n",
    "}\n",
    "\n",
    "#' Create a 6-value, 24-column key for plates\n",
    "#'\n",
    "#' Create a 24-column key with 6 values repeated over 24 plate columns.\n",
    "#' Each of the 6 values is repeated over 3x +RT Techreps and 1x -RT.\n",
    "#' \n",
    "#' This helps to create plate layouts with standard designs.\n",
    "#'\n",
    "#' @param ... Vectors of length 6 describing well contents,\n",
    "#' e.g. sample_id or target_id\n",
    "#' @return tibble (data frame) with 24 rows, and columns\n",
    "#' well_col, prep_type, tech_rep, and supplied values.\n",
    "#'\n",
    "#' @examples\n",
    "#' create_colkey_6_in_24(sample_id=LETTERS[1:6])\n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tibble tibble as_tibble\n",
    "\n",
    "#' @importFrom tidyr %>%\n",
    "#'\n",
    "create_colkey_6_in_24 <- function(...) {\n",
    "    colkey <- tibble(well_col  = factor(1:24),\n",
    "                     prep_type = as_factor(c(rep(\"+RT\", 18), rep(\"-RT\", 6))),\n",
    "                     tech_rep  = as_factor(rep(c(1, 2, 3, 1), each = 6))\n",
    "                     )\n",
    "    if (!missing(...)) {\n",
    "        pieces6 <- list(...) %>% as_tibble()\n",
    "        assertthat::assert_that(nrow(pieces6) == 6, \n",
    "                                msg = \"Some input data is not of length 6\")\n",
    "        pieces24 <- dplyr::bind_rows(pieces6, pieces6, pieces6, pieces6)\n",
    "        colkey <- dplyr::bind_cols(colkey, pieces24)\n",
    "    }\n",
    "    return(colkey)\n",
    "}\n",
    "\n",
    "#' Create a 4-dilution column key for primer calibration\n",
    "#'\n",
    "#' Creates a 24-column key for primer calibration, with 2x biol_reps and 2x\n",
    "#' tech_reps, and 5-fold dilution until 5^4 of +RT; then -RT (no reverse\n",
    "#' transcriptase), NT (no template) negative controls. That is a total of 6\n",
    "#' versions of each sample replicate.\n",
    "#'\n",
    "#' @param dilution Numeric vector of length 6 describing sample dilutions\n",
    "#' @param dilution_nice Character vector of length 6 with nice labels for sample\n",
    "#'   dilutions\n",
    "#' @param prep_type Character vector of length 6 describing type of sample (+RT,\n",
    "#'   -RT, NT)\n",
    "#' @param biol_rep Character vector of length 6 describing biological replicates\n",
    "#' @param tech_rep Character vector of length 6 describing technical replicates\n",
    "#' @return tibble (data frame) with 24 rows, and columns well_col, dilution,\n",
    "#'   dilution_nice, prep_type, biol_rep, tech_rep.\n",
    "#' @examples\n",
    "#' create_colkey_4diln_2ctrl_in_24()\n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tibble tibble\n",
    "#' @importFrom forcats as_factor\n",
    "#'\n",
    "create_colkey_4diln_2ctrl_in_24 <- function(\n",
    "                     dilution      = c(5 ^ (0:-3), 1, 1),\n",
    "                     dilution_nice = c(\"1x\", \"5x\", \"25x\", \"125x\", \"-RT\", \"NT\"),\n",
    "                     prep_type     = c(rep(\"+RT\", 4), \"-RT\", \"NT\"),\n",
    "                     biol_rep      = rep(c(\"A\", \"B\"), each = 12,\n",
    "                                        length.out = 24),\n",
    "                     tech_rep      = rep(1:2, each = 6,\n",
    "                                        length.out = 24)\n",
    "                     ) {\n",
    "    tibble(well_col = factor(1:24),\n",
    "           dilution = rep(dilution, 4),\n",
    "           dilution_nice = rep(dilution_nice, 4),\n",
    "           prep_type = as_factor(rep(prep_type, 4)),\n",
    "           biol_rep = as_factor(biol_rep),\n",
    "           tech_rep = as_factor(tech_rep)\n",
    "    )\n",
    "}\n",
    "\n",
    "#' Create a 6-dilution column key for primer calibration\n",
    "#'\n",
    "#' Creates a 24-column key for primer calibration, with 1x biol_reps and 3x\n",
    "#' tech_reps, and 5-fold dilution until 5^6 of +RT; then -RT (no reverse\n",
    "#' transcriptase), NT (no template) negative controls. That is a total of 8\n",
    "#' versions of each replicate.\n",
    "#'\n",
    "#' @param dilution Numeric vector of length 8 describing sample dilutions\n",
    "#' @param dilution_nice Character vector of length 8 with nice labels for sample\n",
    "#'   dilutions\n",
    "#' @param prep_type Character vector of length 8 describing type of sample (+RT,\n",
    "#'   -RT, NT)\n",
    "#' @param tech_rep Character vector of length 8 describing technical replicates\n",
    "#' @return tibble (data frame) with 24 rows, and variables well_col, dilution,\n",
    "#'   dilution_nice, prep_type, biol_rep, tech_rep.\n",
    "#' @examples\n",
    "#' create_colkey_6diln_2ctrl_in_24()\n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tibble tibble\n",
    "#' @importFrom forcats as_factor\n",
    "#'\n",
    "create_colkey_6diln_2ctrl_in_24 <- function(\n",
    "                     dilution = c(5 ^ (0:-5), 1, 1),\n",
    "                     dilution_nice = c(\"1x\", \"5x\", \"25x\", \"125x\",\n",
    "                                    \"625x\", \"3125x\", \"-RT\", \"NT\"),\n",
    "                     prep_type=c(rep(\"+RT\", 6), \"-RT\", \"NT\"),\n",
    "                     tech_rep = rep(1:3, each = 8, length.out = 24)\n",
    "                     ) {\n",
    "    tibble(well_col = factor(1:24),\n",
    "           dilution = rep(dilution, 3),\n",
    "           dilution_nice = rep(dilution_nice, 3),\n",
    "           prep_type = as_factor(rep(prep_type, 3)),\n",
    "           tech_rep = as_factor(tech_rep))\n",
    "}\n",
    "\n",
    "#' Create a 4-value, 16-row key for plates\n",
    "#'\n",
    "#' Create a 16-row key with 4 values repeated over 16 plate rows. Each of the 4\n",
    "#' values is repeated over 3x +RT Techreps and 1x -RT.\n",
    "#' \n",
    "#' This helps to create plate layouts with standard designs.\n",
    "#'\n",
    "#' @param ... Vectors of length 4 describing well contents, e.g. sample_id or\n",
    "#'   target_id\n",
    "#' @return tibble (data frame) with 16 rows, and variables well_row, prep_type,\n",
    "#'   tech_rep, and supplied values.\n",
    "#' @examples\n",
    "#' create_rowkey_4_in_16(sample_id=c(\"sheep\",\"goat\",\"cow\",\"chicken\"))\n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tibble tibble as_tibble\n",
    "#' @importFrom forcats as_factor\n",
    "#' @importFrom tidyr %>%\n",
    "#'\n",
    "create_rowkey_4_in_16 <- function(...) {\n",
    "    rowkey <- tibble(well_row = factor(LETTERS[1:16]),\n",
    "                     prep_type = as_factor(c(rep(\"+RT\", 12), rep(\"-RT\", 4))),\n",
    "                     tech_rep = as_factor(rep(c(1, 2, 3, 1), each = 4))\n",
    "    )\n",
    "    if (!missing(...)) {\n",
    "        pieces4 <- list(...) %>% as_tibble()\n",
    "        assertthat::assert_that(nrow(pieces4) == 4, \n",
    "                                msg = \"Some input data is not of length 4\")\n",
    "        pieces16 <- dplyr::bind_rows(pieces4, pieces4, pieces4, pieces4)\n",
    "        rowkey <- dplyr::bind_cols(rowkey, pieces16)\n",
    "    }\n",
    "    return(rowkey)\n",
    "}\n",
    "\n",
    "#' Create a plain 8-value, 16-row key for plates\n",
    "#'\n",
    "#' Create a 16-row key with 8 values repeated over 16 plate rows. No other\n",
    "#' information is included by default, hence \"plain\".\n",
    "#' \n",
    "#' This helps to create plate layouts with standard designs.\n",
    "#'\n",
    "#' @param ... Vectors of length 8 describing well contents, e.g. sample or\n",
    "#'   probe.\n",
    "#' @return tibble (data frame) with 16 rows, and variables well_col, and\n",
    "#'   supplied values.\n",
    "#' @examples\n",
    "#' create_rowkey_8_in_16_plain(sample_id=c(\"me\",\"you\",\"them\",\"him\",\n",
    "#'                                    \"her\",\"dog\",\"cat\",\"monkey\"))\n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom tibble tibble\n",
    "#' @importFrom tidyr %>%\n",
    "#'\n",
    "create_rowkey_8_in_16_plain <- function(...) {\n",
    "    rowkey <- tibble(well_row = factor(LETTERS[1:16]))\n",
    "    if (!missing(...)) {\n",
    "        pieces8 <- list(...) %>% as_tibble()\n",
    "        assertthat::assert_that(nrow(pieces8) == 8, \n",
    "                                msg = \"Some input data is not of length 8\")\n",
    "        pieces16 <- dplyr::bind_rows(pieces8, pieces8)\n",
    "        rowkey <- dplyr::bind_cols(rowkey, pieces16)\n",
    "    }\n",
    "    return(rowkey)\n",
    "}\n",
    "\n",
    "#' Label a plate with sample and probe information\n",
    "#'\n",
    "#' For more help, examples and explanations, see the plate setup vignette:\n",
    "#' \\code{vignette(\"platesetup_vignette\", package = \"tidyqpcr\")}\n",
    "#' \n",
    "#' For worked examples of tidyqpcr analysis with 384-well plates, see:\n",
    "#' \\code{vignette(\"calibration_vignette\", package = \"tidyqpcr\")}\n",
    "#'\n",
    "#' @param plate tibble (data frame) with variables well_row, well_col, well.\n",
    "#'   This would usually be produced by create_blank_plate(). It is possible to\n",
    "#'   include other information in additional variables.\n",
    "#' @param rowkey tibble (data frame) describing plate rows, with variables\n",
    "#'   well_row and others.\n",
    "#' @param colkey tibble (data frame) describing plate columns, with variables\n",
    "#'   well_col and others.\n",
    "#' @param coercefactors if TRUE, coerce well_row in rowkey and well_col in\n",
    "#'   colkey to factors\n",
    "#'\n",
    "#' @return tibble (data frame) with variables well_row, well_col, well, and\n",
    "#'   others.\n",
    "#'\n",
    "#'   This tibble contains all combinations of well_row and well_col found in the\n",
    "#'   input plate, and all information supplied in rowkey and colkey distributed\n",
    "#'   across every well of the plate. Return plate is ordered by row well_row\n",
    "#'   then column well_col.\n",
    "#'\n",
    "#'   Note this ordering may cause a problem if well_col is supplied as a\n",
    "#'   character (1,10,11,...), instead of a factor or integer (1,2,3,...). For\n",
    "#'   this reason, the function by default converts well_row in `rowkey`, and\n",
    "#'   well_col in `colkey`, to factors, taking factor levels from `plate`, and\n",
    "#'   messages the user.\n",
    "#'   \n",
    "#'   If `plate$well_col` or `plate$well_row` are not factors and coercefactors = TRUE \n",
    "#'   label_plate_rowcol will automatically convert them to factors, but will output a \n",
    "#'   warning telling users this may lead to unexpected behaviour. \n",
    "#'\n",
    "#'   Other tidyqpcr functions require plate plans to contain variables\n",
    "#'   sample_id, target_id, and prep_type, so `label_plate_rowcol` will message\n",
    "#'   if any of these are missing. This is a message, not an error, because these\n",
    "#'   variables can be added by users later.\n",
    "#'\n",
    "#' @examples\n",
    "#' label_plate_rowcol(plate = create_blank_plate()) # returns blank plate\n",
    "#' \n",
    "#' # label blank 96-well plate with empty edge wells\n",
    "#' \n",
    "#' label_plate_rowcol(plate = create_blank_plate(well_row = LETTERS[2:7], \n",
    "#'                                               well_col = 2:11))\n",
    "#' \n",
    "#' # label 96-well plate with sample id in rows\n",
    "#' \n",
    "#' label_plate_rowcol(plate = create_blank_plate(well_row = LETTERS[1:8],\n",
    "#'                                               well_col = 1:12),\n",
    "#'                    rowkey = tibble(well_row = LETTERS[1:8],\n",
    "#'                                    sample_id = paste0(\"S_\",1:8)))\n",
    "#' \n",
    "#' # label fraction of 96-well plate with target id in columns\n",
    "#' \n",
    "#' label_plate_rowcol(plate = create_blank_plate(well_row = LETTERS[1:8],\n",
    "#'                                               well_col = 1:4),\n",
    "#'                    colkey = tibble(well_col = 1:4,\n",
    "#'                                    target_id = paste0(\"T_\",1:4)))\n",
    "#' \n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#'\n",
    "#' @importFrom rlang .data\n",
    "#' @importFrom forcats as_factor\n",
    "#'\n",
    "label_plate_rowcol <- function(plate,\n",
    "                               rowkey = NULL,\n",
    "                               colkey = NULL,\n",
    "                               coercefactors = TRUE) {\n",
    "    assertthat::assert_that(\n",
    "        assertthat::has_name(plate, \n",
    "                             c(\"well_row\",\"well_col\")))\n",
    "    \n",
    "    if (!is.factor(plate$well_col) & coercefactors){\n",
    "        warning(\"plate$well_col is not a factor. Automatically generating plate$well_col factor levels. May lead to incorrect plate plans.\")\n",
    "        plate <- plate %>%\n",
    "            dplyr::mutate(well_col = as_factor(.data$well_col))\n",
    "    }\n",
    "    \n",
    "    if (!is.factor(plate$well_row) & coercefactors){\n",
    "        warning(\"plate$well_row is not a factor. Automatically generating plate$well_row factor levels. May lead to incorrect plate plans.\")\n",
    "        plate <- plate %>%\n",
    "            dplyr::mutate(well_row = as_factor(.data$well_row))\n",
    "    }\n",
    "    \n",
    "    if (!is.null(colkey)) {\n",
    "        assertthat::assert_that(assertthat::has_name(colkey, \"well_col\"))\n",
    "        # Note: should this if clause be a freestanding function?\n",
    "        # coerce_column_to_factor(df, col, warn=FALSE)?\n",
    "        if (!is.factor(colkey$well_col) & coercefactors) {\n",
    "            message(\"coercing well_col to a factor with levels from plate$well_col\")\n",
    "            colkey <- dplyr::mutate(\n",
    "                colkey,\n",
    "                well_col = factor(.data$well_col,\n",
    "                                  levels = levels(plate$well_col))\n",
    "            )\n",
    "        }\n",
    "        plate <- dplyr::left_join(plate, colkey, by = \"well_col\")\n",
    "    }\n",
    "    if (!is.null(rowkey)) {\n",
    "        assertthat::assert_that(assertthat::has_name(rowkey, \"well_row\"))\n",
    "        if (!is.factor(rowkey$well_row) & coercefactors) {\n",
    "            message(\"coercing well_row to a factor with levels from plate$well_row\")\n",
    "            rowkey <- dplyr::mutate(\n",
    "                rowkey,\n",
    "                well_row = factor(.data$well_row,\n",
    "                                  levels = levels(plate$well_row))\n",
    "            )\n",
    "        }\n",
    "        plate <- dplyr::left_join(plate, rowkey, by = \"well_row\")\n",
    "    }\n",
    "    # check that plate contains sample_id, target_id, prep_type\n",
    "    if (! \"sample_id\" %in% names(plate)) {\n",
    "        message(\"plate does not contain variable sample_id\")\n",
    "    }\n",
    "    if (! \"target_id\" %in% names(plate)) {\n",
    "        message(\"plate does not have variable target_id\")\n",
    "    }\n",
    "    if (! \"prep_type\" %in% names(plate)) {\n",
    "        message(\"plate does not have variable prep_type\")\n",
    "    }\n",
    "    return(dplyr::arrange(plate, .data$well_row, .data$well_col))\n",
    "}\n",
    "\n",
    "\n",
    "#' Display an empty plate plan which can be populated with \n",
    "#' ggplot2 geom elements.\n",
    "#'\n",
    "#' @param plate tibble with variables well_col, well_row.\n",
    "#'\n",
    "#' @return ggplot object; major output is to plot it\n",
    "#'\n",
    "#' @examples\n",
    "#' library(ggplot2)\n",
    "#' \n",
    "#' # display empty plot of empty plate\n",
    "#' display_plate(create_blank_plate_96well())\n",
    "#' \n",
    "#' # display wells of empty plate filled by column\n",
    "#' display_plate(create_blank_plate_96well()) + \n",
    "#'   geom_tile(aes(fill = well_col), colour = \"black\")\n",
    "#' \n",
    "#' # display wells of empty 1536-well plate filled by row\n",
    "#' display_plate(create_blank_plate_1536well()) + \n",
    "#'   geom_tile(aes(fill = well_row), colour = \"black\")\n",
    "#' \n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom forcats as_factor\n",
    "#' @importFrom rlang .data\n",
    "#'\n",
    "display_plate <- function(plate) {\n",
    "    assertthat::assert_that(\n",
    "        assertthat::has_name(plate, \n",
    "                             c(\"well_row\",\"well_col\")))\n",
    "    \n",
    "    rowlevels <- \n",
    "        dplyr::pull(plate, .data$well_row) %>%\n",
    "        as_factor() %>%\n",
    "        levels()\n",
    "\n",
    "    ggplot2::ggplot(data = plate,\n",
    "                    ggplot2::aes(x = as_factor(.data$well_col),\n",
    "                        y = as_factor(.data$well_row))) +\n",
    "        ggplot2::scale_x_discrete(expand = c(0, 0)) +\n",
    "        ggplot2::scale_y_discrete(expand = c(0, 0),\n",
    "                                  limits = rev(rowlevels)) +\n",
    "        ggplot2::coord_equal() +\n",
    "        ggplot2::theme_void() +\n",
    "        ggplot2::theme(axis.text = ggplot2::element_text(angle = 0),\n",
    "                       panel.grid.major = ggplot2::element_blank(),\n",
    "                       legend.position = \"none\",\n",
    "                       plot.margin = grid::unit(rep(0.01, 4), \"npc\"),\n",
    "                       panel.border = ggplot2::element_blank())\n",
    "}\n",
    "\n",
    "#' Display qPCR plate plan with sample_id, target_id, prep_type per well\n",
    "#'\n",
    "#' @param plate tibble with variables well_col, well_row, sample_id, target_id,\n",
    "#'   prep_type. Output from label_plate_rowcol.\n",
    "#'\n",
    "#' @return ggplot object; major output is to plot it\n",
    "#'\n",
    "#' @examples \n",
    "#' \n",
    "#' \n",
    "#' # create basic 6-well plate\n",
    "#' basic_plate <- \n",
    "#'     label_plate_rowcol(plate = create_blank_plate(well_row = LETTERS[1:2],\n",
    "#'                                                   well_col = 1:3),\n",
    "#'                        rowkey = tibble(well_row = factor(LETTERS[1:2]),\n",
    "#'                                        target_id = c(\"T_A\",\"T_B\")),\n",
    "#'                        colkey = tibble(well_col = factor(1:3),\n",
    "#'                                        sample_id = c(\"S_1\",\"S_2\", \"S_3\"),\n",
    "#'                                        prep_type = \"+RT\"))\n",
    "#' \n",
    "#' # display basic plate\n",
    "#' display_plate_qpcr(basic_plate)\n",
    "#' \n",
    "#' # create full 384 well plate\n",
    "#' full_plate <- label_plate_rowcol(create_blank_plate(), \n",
    "#'                                   create_rowkey_8_in_16_plain(target_id = c(\"T_1\", \"T_2\",\n",
    "#'                                                                             \"T_3\", \"T_4\", \n",
    "#'                                                                             \"T_5\", \"T_6\",\n",
    "#'                                                                             \"T_7\", \"T_8\")), \n",
    "#'                                   create_colkey_6diln_2ctrl_in_24() %>% \n",
    "#'                                       dplyr::mutate(sample_id = paste0(dilution_nice,\n",
    "#'                                                                        \"_\",\n",
    "#'                                                                        tech_rep)))\n",
    "#' \n",
    "#' # display full plate\n",
    "#' display_plate_qpcr(full_plate)\n",
    "#' \n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom rlang .data\n",
    "#'\n",
    "display_plate_qpcr <- function(plate) {\n",
    "    assertthat::assert_that(\n",
    "        assertthat::has_name(plate, \n",
    "                             c(\"target_id\",\n",
    "                               \"sample_id\",\n",
    "                               \"prep_type\")))\n",
    "    \n",
    "    display_plate(plate) +\n",
    "        ggplot2::geom_tile(ggplot2::aes(fill = .data$target_id), \n",
    "                           alpha = 0.3) +\n",
    "        ggplot2::geom_text(ggplot2::aes(label = \n",
    "                                            paste(.data$target_id,\n",
    "                                                  .data$sample_id,\n",
    "                                                  .data$prep_type,\n",
    "                                                  sep = \"\\n\")),\n",
    "                           size = 2.5, lineheight = 1)\n",
    "}\n",
    "\n",
    "#' Display the value of each well across the plate. \n",
    "#' \n",
    "#' Plots the plate with each well coloured by its value. Example values are Cq, Delta Cq or Delta Delta Cq.\n",
    "#'\n",
    "#' For a specific example see the calibration vignette:\n",
    "#' \\code{vignette(\"calibration_vignette\", package = \"tidyqpcr\")}\n",
    "#'\n",
    "#' @param plate tibble with variables well_col, well_row, and the variable to be plotted.\n",
    "#' \n",
    "#' @param value character vector selecting the variable in plate to plot as the well value\n",
    "#'\n",
    "#' @return ggplot object; major output is to plot it\n",
    "#'\n",
    "#' @examples \n",
    "#' library(dplyr)\n",
    "#' library(ggplot2)\n",
    "#' \n",
    "#' # create 96 well plate with random values\n",
    "#' plate_randomcq <- create_blank_plate_96well() %>%\n",
    "#'     mutate(cq = runif(96) * 10,\n",
    "#'            deltacq = runif(96) * 2)\n",
    "#' \n",
    "#' \n",
    "#' # display well Cq value across plate\n",
    "#' display_plate_value(plate_randomcq)\n",
    "#' \n",
    "#' # display well Delta Cq value across plate with red colour pallette\n",
    "#' display_plate_value(plate_randomcq, value = \"deltacq\") +   # uses ggplot syntax\n",
    "#'     scale_fill_gradient(high = \"#FF0000\") \n",
    "#'           \n",
    "#' \n",
    "#' @family plate creation functions\n",
    "#'\n",
    "#' @export\n",
    "#' @importFrom forcats as_factor\n",
    "#' @importFrom rlang .data\n",
    "#'\n",
    "display_plate_value <- function(plate, value = \"cq\") {\n",
    "    # check value exists in given plate\n",
    "    assertthat::assert_that(value %in% names(plate), \n",
    "                            msg = paste0(value, \" is not the name of a variable in the given plate\"))\n",
    "    \n",
    "    # check each well has one value only\n",
    "    unique_well_value <- plate %>%\n",
    "        dplyr::group_by(.data$well) %>%\n",
    "        dplyr::summarise(num_well = dplyr::n()) %>%\n",
    "        dplyr::mutate(not_equal_one = .data$num_well != 1)\n",
    "    \n",
    "    assertthat::assert_that(sum(unique_well_value$not_equal_one) == 0, \n",
    "                            msg = paste0(\"Wells do not have unique \", value, \" value.\"))\n",
    "    \n",
    "    rowlevels <- \n",
    "        dplyr::pull(plate, .data$well_row) %>%\n",
    "        as_factor() %>%\n",
    "        levels()\n",
    "    \n",
    "    display_plate(plate = plate) +\n",
    "        ggplot2::geom_tile(ggplot2::aes(fill = .data[[value]])) +\n",
    "        ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5),\n",
    "                       panel.grid.major = ggplot2::element_blank()) +\n",
    "        ggplot2::labs(title = paste0({{value}}, \" values for each well across the plate\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ee5eddb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Reads raw text-format fluorescence data in 1 colour from Roche Lightcyclers\n",
    "#' \n",
    "#' This is the data from \"export in text format\" from the Lightcycler software.\n",
    "#' The other data format, .ixo, can be converted to .txt format by the\n",
    "#' Lightcycler software.\n",
    "#'\n",
    "#' This function is a thin wrapper around readr::read_tsv.\n",
    "#'\n",
    "#' @param filename file name\n",
    "#' @param skip number of lines to skip, defaults to 2\n",
    "#' @param col_names names to give to columns\n",
    "#' @param col_types data types of columns\n",
    "#' @param ... other arguments to pass to read_tsv, if needed\n",
    "#'\n",
    "#' @return tibble containing raw data, with default column names:\n",
    "#'\n",
    "#'   well: the well of the plate, e.g. A1\n",
    "#'\n",
    "#'   sample_info: this is the \"Sample\" field entered in lightcycler software,\n",
    "#'   defaults to \"Sample X\"\n",
    "#'\n",
    "#'   program_no: the number of the cycler program, for 2-step PCR defaults to 1\n",
    "#'   = melt, 2 = amplify, 3 = melt.\n",
    "#'\n",
    "#'   segment_no: the number of the segment of the cycler program, e.g.\n",
    "#'   hold/raise/lower temperature\n",
    "#'\n",
    "#'   cycle: the cycle number, for programs with repeated cycles (i.e.\n",
    "#'   amplification)\n",
    "#'\n",
    "#'   time: the time of fluorescence reading acquisition (in what units???)\n",
    "#'\n",
    "#'   temperature: the temperature of the block at fluorescence acquisition\n",
    "#'\n",
    "#'   fluor_raw: the raw fluorescence reading in \"arbitrary units\". For SYBR safe, this\n",
    "#'   would be 483nm excitation, 533nm emission.\n",
    "#'   \n",
    "#' @export\n",
    "#' @seealso read_lightcycler_1colour_cq\n",
    "#'\n",
    "#' @examples read_lightcycler_1colour_raw(system.file(\"extdata/Edward_qPCR_Nrd1_calibration_2019-02-02.txt.gz\", \n",
    "#'                                                  package = \"tidyqpcr\"))\n",
    "#'\n",
    "read_lightcycler_1colour_raw <- function(\n",
    "    filename,\n",
    "    skip = 2,\n",
    "    col_names = c(\n",
    "        \"well\", \"sample_info\", \"program_no\", \"segment_no\",\n",
    "        \"cycle\", \"time\", \"temperature\", \"fluor_raw\"\n",
    "    ), \n",
    "    col_types = \"ccffinnn\",\n",
    "    ...) {\n",
    "    readr::read_tsv(file = filename,\n",
    "                    skip = skip,\n",
    "                    col_names = col_names,\n",
    "                    col_types = col_types,\n",
    "                    ...)\n",
    "}\n",
    "\n",
    "#' Reads quantification cycle (cq) data in 1 colour from Roche Lightcyclers\n",
    "#'\n",
    "#' This is the data from \"export in text format\" from the analysis tab in the\n",
    "#' Lightcycler software. That software calls cq, \"Cp\".\n",
    "#'\n",
    "#' This function is a thin wrapper around readr::read_tsv.\n",
    "#'\n",
    "#' @param filename file name\n",
    "#' @param skip number of lines to skip, defaults to 2\n",
    "#' @param col_names names to give to columns\n",
    "#' @param col_types data types of columns\n",
    "#' @param ... other arguments to pass to read_tsv, if needed\n",
    "#'\n",
    "#' @return tibble containing cq data\n",
    "#' @export\n",
    "#' @seealso read_lightcycler_1colour_raw\n",
    "#'\n",
    "#' @examples read_lightcycler_1colour_cq(system.file(\"extdata/Edward_qPCR_Nrd1_calibration_2019-02-02_Cq.txt.gz\", \n",
    "#'                                                  package = \"tidyqpcr\"))\n",
    "#'\n",
    "read_lightcycler_1colour_cq <- function(\n",
    "    filename,\n",
    "    skip = 2,\n",
    "    col_names = c(\n",
    "        \"include\", \"color\", \"well\", \"sample_info\",\n",
    "        \"cq\", \"concentration\", \"standard\", \"status\"\n",
    "    ), \n",
    "    col_types = \"liccddil\",\n",
    "    ...) {\n",
    "    readr::read_tsv(file = filename,\n",
    "                    skip = 2,\n",
    "                    col_names=col_names,\n",
    "                    col_types=col_types,\n",
    "                    ...)\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
